### DISCUSSION TITLE
Allow return type on atom setters

### DISCUSSION THREAD

**Discussion by carlesba** [pmndrs/jotai#1827] [Ideas] [answered]
Currently, atom setters don't return any value. This is especially frustrating when setters run async tasks as prevents event handlers to manage side effects.

For instance:
```jsx

const mutateAtom = atom(null, async (get, set, payload)=> {
  const value = await fetch(payload)
  // here we can handle side effects in Atom's land
  set(otherAtom, value)
  return value // this won't don anything
})

const Component =()=> {
  const [, mutate] = useAtom(mutateAtom)
  return <div
    onClick={()=> {
       mutate({value}).then(result=> {
           /* 
this is not possible at the moment as setters don't return values
We cannot handle effects in components land, we're forced to move them into Atoms
But some side effects don't belong to Atoms and need to be handled in UI  
*/
           notifySuccess() 
       })
   }}
   /> 


```

A work around here is to allow a **callback** function in setter options so caller can be notified about the result.

```ts
const mutation = atom(null, (get, set, payload:  {value: string, callback: ()=> void})=> {
   fetch(payload.value).then(callback)
})
```

I find this not very handy as setter's payloads work very well describing event's requirements. Callbacks here are just process interfaces to help orchestrating a flow and should be handled at a different level.

I'm missing jotai's internal implementation and I'm aware this may not be possible. However, I think it would improve developer experience and atom's composability.

**Comment by dai-shi** [comment, marked as answer]
> atom setters don't return any value.

Are you sure? v1 didn't do it, but v2 should do.

**Reply by carlesba** [reply]
My bad! Haven't time to check v2 yet. That will be amazing. Thank you very much!
